ci/eval/compare: add 10.rebuild-nixos-tests label

wolfgangwalther opened the pull request:
This indicates that the NixOS test-driver changed and all NixOS tests have to be rebuilt. It can be used to either re-target to staging or to batch this with other similar changes, at least.

I created the new label with the following description for now:

> This PR causes rebuilds for all NixOS tests and should normally target the staging branches.

*(before anyone wants to suggest changes, this has a 100 character limit, keep that in mind!)*

This can later be adjusted, when the `kernel-updates-rebuilding-all-nixos-tests` branch exists.

Resolves #437208

## Things done

- [x] Tested in my fork.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
ci/eval: eval nixosTests.simple

We can't eval all nixos tests, this would be way too expensive. But we
can evaluate `nixosTests.simple` as an indication whether the nixos test
driver has changed. If that's the case, this means that *all* NixOS
tests need to be rebuilt.

wolfgangwalther committed:
ci/eval/compare: add 10.rebuild-nixos-tests label

This indicates that the NixOS test-driver changed and all NixOS tests
have to be rebuilt. It can be used to either re-target to staging or to
batch this with other similar changes, at least.

emilazy commented on ci/eval/outpaths.nix:
```diff
@@ -90,4 +94,9 @@ let
+tweak (
+  (builtins.removeAttrs nixpkgsJobs blacklist)
+  // {
+    nixosTests.simple = nixosJobs.tests.simple;
```
If we use `simple` as the exemplar for this, can we drop the minimal profile from it? The majority of tests do not use it, and I think that we should catch changes to things that would rebuild all tests not using the minimal profile.

OTOH, this would cause the `simple` test to involve more builds itself, which might go against using it as a NixOS smoke test for mass rebuilds? Maybe we could split it out into two tests or something, but that sounds kind of annoying.

LordGrimmauld replied:
I would +1 a real "neutral" test, whether that is `simple` actually made simple or something else. Something that does basically just `machine.succeed("echo 'hello world'")`.

The `simple` test is certainly not sufficient to smoke-test systemd. It is decent enough for smoke-testing qemu itself, but for that any test would work. I personally don't see a real need for a minimal test, but if anyone would *want* a minimal test, we can move the current `simple` test to be named `minimal` (or similar, indicating its intended to be small) and create a new (more neutral) `simple` test.

We have `nixosTests.systemd` as a systemd smoke test. There is some more things going on there, but it should be run on any change to systemd anyways, rebuilds be damned. I at least wouldn't feel confident opening/merging any PR without running that.

emilazy replied:
> Now, if the scope is "catching cases where most NixOS tests rebuild, independently whether caused by basic NixOS config changes or test driver", then we should remove the minimal profile, because we should keep the test as "default" as possible.

Right. I think this is indeed what we want to catch, because it’s the end result on Hydra that matters, not what component they happen to touch. (Any given kernel version won’t rebuild *every* test, either, because some of them use other kernels.)

I guess re‐reading the commit message there’s nothing in it I think is incorrect; I just got the impression that the reason to desire it hadn’t been fully communicated :)

emilazy commented on ci/eval/outpaths.nix:
```diff
@@ -59,9 +60,15 @@ let
+
+  nixosJobs = import (path + "/nixos/release.nix") {
+    inherit attrNamesOnly;
+    supportedSystems = if systems == null then [ builtins.currentSystem ] else systems;
```
Not sure if there is a point to evaluating/building NixOS tests on Darwin – it is the same NixOS systems, just with the test driver running on Darwin, and that configuration is not built by Hydra. I suppose catching eval regressions there has some value, but I’d likely err on the side of saving the system evals.

emilazy reviewed (commented):
Thank you for working on this 💜

Yes, mentioning `staging` in the label sounds okay to me until we get a branch. It doesn’t hurt to send such changes to `staging` and kernel maintainers will know it doesn’t apply to them.

BTW, if you want GitHub to display the diff of stacked PRs more sensibly, you can push the first PR to a branch in the Nixpkgs repository, and then open a second PR that targets that PR’s branch. This will show only the additional commits compared to the first PR, and when the first PR is merged, the second PR will automatically adjust to target the first PR’s base branch instead. GitHub doesn’t have great support for this workflow, and it’s unfortunate that it requires the preemptive use of branches on the main repo, but it’s a better review experience than the commits being included in both PRs.

wolfgangwalther committed:
nixosTests.simple: drop minimal profile

This should be a test that will change whenever most likely *all* NixOS
tests change. As such it should use the most "default" configuration
possible. This decreases chances of a change that will affect almost
other NixOS tests to go unnoticed by this test.

With the minimal profile used previously, there could be changes made to
the default behavior of NixOS modules which were explicitly disabled in
this profile. These would not change the simple test, but would change
all other tests.

Removing as many moving parts as possible from this test should make it
the most default it can be. It now contains only pieces that every other
NixOS test has, too: A name, a machine and a testScript. But nothing
fancy.

Ma27 reviewed (commented):
One suggestion (I won't die on that hill, though).
Otherwise, looks reasonable, thanks!

MattSturgeon reviewed (commented):
Just commenting to say that I see no issues in the diff, but I don't feel confident enough in these weeds to sign off with an approval. I trust those more familiar will do so.

philiptaron reviewed (approved):
I agree that the way `ci/eval` and the various release top-level derivation entry points fit together is pretty awkward. Fixing that or even trying to find a direction to resolve that awkwardness doesn't fit with this scope of this PR.

So, I support merging this only minor tweaks and continuing to evolve the interface in refactor PRs.

LordGrimmauld commented:
Yep, tag is there, seems to work just fine. Nice!
